File: EventServiceNodeJS/src/EventHandler.test.ts
import { EventHandler } from './EventHandler';
import { EventMessageInstance, EventMessageType } from './EventMessageInstance';

const eventMessageInstances = [
    new EventMessageInstance(
        '01',
        EventMessageType.REQUEST,
        'GetDataUsage',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '01',
        EventMessageType.RESPONSE,
        'GetDataUsage',
        1725392155,
        true
    ),
    new EventMessageInstance(
        '021',
        EventMessageType.REQUEST,
        'Action',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '022',
        EventMessageType.RESPONSE,
        'Action',
        1725392155,
        false
    ),
    new EventMessageInstance(
        '031',
        EventMessageType.REQUEST,
        'GetDataUsage',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '032',
        EventMessageType.RESPONSE,
        'GetDataUsage',
        1725392155,
        false
    ),
    new EventMessageInstance(
        '041',
        EventMessageType.REQUEST,
        'Action',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '041',
        EventMessageType.RESPONSE,
        'Action',
        1725392155,
        true
    ),
    new EventMessageInstance(
        '051',
        EventMessageType.REQUEST,
        'GetDataUsage',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '052',
        EventMessageType.RESPONSE,
        'GetDataUsage',
        1725392155,
        false
    ),
    new EventMessageInstance(
        '061',
        EventMessageType.REQUEST,
        'Action',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '062',
        EventMessageType.RESPONSE,
        'Action',
        1725392155,
        false
    ),
    new EventMessageInstance(
        '071',
        EventMessageType.REQUEST,
        'GetDataUsage',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '071',
        EventMessageType.RESPONSE,
        'GetDataUsage',
        1725392155,
        false
    ),
    new EventMessageInstance(
        '081',
        EventMessageType.REQUEST,
        'GetDataUsage',
        1725392080,
        true
    ),
    new EventMessageInstance(
        '082',
        EventMessageType.RESPONSE,
        'GetDataUsage',
        1725392155,
        false
    )
];

test('Testing getNumberPendingRequests()', () => {
    const eventHandler = new EventHandler();
    expect(eventHandler.getNumberPendingRequests(eventMessageInstances)).toBe(
        5
    );
});

test('Testing GetNumberPendingRequestsByOperationName()', () => {
    const eventHandler = new EventHandler();
    expect(
        eventHandler.getNumberPendingRequestsByOperationName(
            eventMessageInstances,
            'GetDataUsage'
        )
    ).toBe(3);
});

File: EventServiceCSharp/EventProcessor.Test/EventHandler_Test.cs
namespace EventProcessor.Test;

[TestClass]
public class EventHandler_Test
{
    private EventMessageInstance[] eventMessageInstances = [];

    [TestInitialize]
    public void SetUp()
    {
        eventMessageInstances =
        [
            new EventMessageInstance
            {
                TransactionId = "01",
                Type = EventMessageType.REQUEST,
                Operation = "GetDataUsage",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "01",
                Type = EventMessageType.RESPONSE,
                Operation = "GetDataUsage",
                Timestamp = 1725392155,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "021",
                Type = EventMessageType.REQUEST,
                Operation = "Action",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "022",
                Type = EventMessageType.RESPONSE,
                Operation = "Action",
                Timestamp = 1725392155,
                Success = false
            },
            new EventMessageInstance
            {
                TransactionId = "031",
                Type = EventMessageType.REQUEST,
                Operation = "GetDataUsage",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "032",
                Type = EventMessageType.RESPONSE,
                Operation = "GetDataUsage",
                Timestamp = 1725392155,
                Success = false
            },
            new EventMessageInstance
            {
                TransactionId = "041",
                Type = EventMessageType.REQUEST,
                Operation = "Action",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "041",
                Type = EventMessageType.RESPONSE,
                Operation = "Action",
                Timestamp = 1725392155,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "051",
                Type = EventMessageType.REQUEST,
                Operation = "GetDataUsage",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "052",
                Type = EventMessageType.RESPONSE,
                Operation = "GetDataUsage",
                Timestamp = 1725392155,
                Success = false
            },
            new EventMessageInstance
            {
                TransactionId = "061",
                Type = EventMessageType.REQUEST,
                Operation = "Action",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "062",
                Type = EventMessageType.RESPONSE,
                Operation = "Action",
                Timestamp = 1725392155,
                Success = false
            },
            new EventMessageInstance
            {
                TransactionId = "071",
                Type = EventMessageType.REQUEST,
                Operation = "GetDataUsage",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "071",
                Type = EventMessageType.RESPONSE,
                Operation = "GetDataUsage",
                Timestamp = 1725392155,
                Success = false
            },
            new EventMessageInstance
            {
                TransactionId = "081",
                Type = EventMessageType.REQUEST,
                Operation = "GetDataUsage",
                Timestamp = 1725392080,
                Success = true
            },
            new EventMessageInstance
            {
                TransactionId = "082",
                Type = EventMessageType.RESPONSE,
                Operation = "GetDataUsage",
                Timestamp = 1725392155,
                Success = false
            }
        ];
    }

    [TestMethod]
    public void TestGetNumberPendingRequests()
    {
        Console.WriteLine("count: " + eventMessageInstances.Length);
        var eventService = new EventHandler();
        int result = eventService.GetNumberPendingRequests(eventMessageInstances);
        Assert.AreEqual(5, result);
    }

    [TestMethod]
    public void TestGetNumberPendingRequestsByOperationName()
    {
        var eventService = new EventHandler();
        const string operationName = "GetDataUsage";
        int result = eventService.GetNumberPendingRequestsByOperationName(eventMessageInstances, operationName);
        Assert.AreEqual(3, result);
    }
}